keeper/component/dispatcher/query_server.go

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: BUSL-1.1
+//
+// Copyright (C) 2025, NASD Inc. All rights reserved.
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file of this repository and at www.mariadb.com/bsl11.
+//
+// ANY USE OF THE LICENSED WORK IN VIOLATION OF THIS LICENSE WILL AUTOMATICALLY
+// TERMINATE YOUR RIGHTS UNDER THIS LICENSE FOR THE CURRENT AND ALL OTHER
+// VERSIONS OF THE LICENSED WORK.
+//
+// THIS LICENSE DOES NOT GRANT YOU ANY RIGHT IN ANY TRADEMARK OR LOGO OF
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN "AS IS" BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package dispatcher
+
+import (
+	"context"
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	dispatchertypes "orbiter.dev/types/component/dispatcher"
+	"orbiter.dev/types/core"
+)
+
+var _ dispatchertypes.QueryServer = &queryServer{}
+
+type queryServer struct {
+	*Dispatcher
+}
+
+func NewQueryServer(d *Dispatcher) dispatchertypes.QueryServer {
+	return queryServer{Dispatcher: d}
+}
+
+func (q queryServer) DispatchedCounts(
+	ctx context.Context,
+	req *dispatchertypes.QueryDispatchedCountsRequest,
+) (*dispatchertypes.QueryDispatchedCountsResponse, error) {
+	if req == nil {
+		return nil, sdkerrors.ErrInvalidRequest
+	}
+
+	sourceID, err := core.NewCrossChainID(req.SourceProtocolId, req.SourceCounterpartyId)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "error creating source cross-chain ID")
+	}
+
+	destID, err := core.NewCrossChainID(req.DestinationProtocolId, req.DestinationCounterpartyId)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "error creating destination cross-chain ID")
+	}
+
+	if !q.HasDispatchedCounts(ctx, &sourceID, &destID) {
+		return nil, fmt.Errorf(
+			"dispatched counts do not exist for source ID %s and destination ID %s",
+			sourceID.String(),
+			destID.String(),
+		)
+	}
+
+	counts := q.GetDispatchedCounts(ctx, &sourceID, &destID)
+
```

Comment:
this is actually calling `GetDispatchedCounts` twice under the hood - no need to do it on this PR but I will create a ticket to improve this going forward and avoid the duplicate state access.

Reply:
https://linear.app/noblexyz/issue/ENG-501/avoid-duplicate-call-to-getdispatchedcounts-in-query-server